User first refactor

JakubKowalski1997 opened the pull request.

JakubKowalski1997 committed:
User first refactor

Pawlllosss commented on src/main/resources/graphql/schema/user.graphqls:
```diff
@@ -14,10 +14,6 @@ input UserLoginDTO {
 }
 
-extend type Mutation {
-    createUser(user: UserDTO !): User !
```
What do you think about actually inverted approach? That means getting rid of `createVenue` and `createBand` mutations. Instead create `createUserBand(user: UserDTO !, band: BandDTO !)` and `createUserVenue(user: UserDTO !, venue: VenueDTO !)` mutations - in that case we can even consider removing `role` from `UserDTO`.

JakubKowalski1997 replied:
hmm splitting this like this maybe make sense

JakubKowalski1997 replied:
removing role would get us out of String Driven Development

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/band/control/BandService.java:
```diff
@@ -37,11 +37,8 @@ public BandService(BandRepository bandRepository, UserRepository userRepository,
-                });
+    public Optional<Band> createBand(Band band) {
+        userService.createUser(band.getUser());
+        return Optional.of(bandRepository.save(band));
```
According to https://github.com/OrkiZMajorki/backend/pull/6/files?file-filters%5B%5D=.graphqls&file-filters%5B%5D=.java&hide-deleted-files=true#r500539084 maybe we should consider inverting approach. That means create a band, linked to user within the `userService`. It may be based on sended `role`.

JakubKowalski1997 replied:
How then we will pass info about both venue and band?

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/venue/control/VenueService.java:
```diff
@@ -11,20 +11,17 @@
-                    return venueRepository.save(venue);
-                });
+    public Optional<Venue> createVenue(Venue venue) {
+        userService.createUser(venue.getUser());
```
Similar to https://github.com/OrkiZMajorki/backend/pull/6/files?file-filters%5B%5D=.graphqls&file-filters%5B%5D=.java&hide-deleted-files=true#r500541099

JakubKowalski1997 committed:
User second refactor

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/band/entity/Band.java:
```diff
@@ -37,7 +36,7 @@
     private String imageUrl;
 
-    @ManyToMany(cascade = CascadeType.ALL)
+    @ManyToMany(cascade = CascadeType.MERGE)
```
Yeah, `CascadeType.ALL` wasn't a good idea, but do you think `MERGE` here is enough?

JakubKowalski1997 replied:
hibernate was throwing some errors so i changed it

Pawlllosss replied:
When these errors happened? I can create user without problems. I would remain every CascadeType, but without `CascadeType.REMOVE`

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/boundary/graphql/UserMutation.java:
```diff
@@ -1,40 +1,55 @@
         User user = convertToEntity(userDTO);
-        userService.createUser(user);
+        Optional<Band> band = bandService.createBand(bandConverter.convertToEntity(bandDTO));
+        band.ifPresent(value -> userService.createUser(user, value.getId()));
```
I think that GraphQL query/mutation classes like rest controllers should only handle converting to/from DTO and business logic should remain within services. I would move these two lines to a separate `createUserBand(user, band)` method within `UserService`.

I would also modify `createBand` method so it doesn't return `optional`, because we don't need to query user to create a band. I think it may be even possible to create band together with user this way (if we have User-Band mapping), without persisting band with a separate band:

```
user.setBand(band);
userRepository.save(user);
```

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/boundary/graphql/UserMutation.java:
```diff
@@ -1,40 +1,55 @@
+    public UserResponseDTO createUserVenue(UserDTO userDTO, VenueDTO venueDTO) {
+        User user = convertToEntity(userDTO);
+        Optional<Venue> venue = venueService.createVenue(venueConverter.convertToEntity(venueDTO));
+        venue.ifPresent(value -> userService.createUser(user, value.getId()));
```
Similar to https://github.com/OrkiZMajorki/backend/pull/6/files#r501274875

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/entity/User.java:
```diff
@@ -26,10 +25,12 @@
     private Role role;
 
+    @Column(name = "role_id")
+    private Long roleId;
```
If we want to store Role as a separate entity on the database (however I am not sure if it's necessary and enum may be enough for now), then in Hibernate we need to create one to one mapping with `Role` object, not it's id - joining based on id will be handled by Hibernate.

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/venue/control/VenueService.java:
```diff
@@ -11,20 +11,16 @@
-                    return venueRepository.save(venue);
-                });
+    public Optional<Venue> createVenue(Venue venue) {
+        return Optional.of(venueRepository.save(venue));
```
I think it's not needed to wrap `Venue` with `Optional` - there's no way that venue won't be created, because of not finding some data.

Pawlllosss commented on src/main/resources/graphql/schema/venue.graphqls:
```diff
@@ -5,15 +5,13 @@ type Venue {
-
-extend type Mutation {
-    createVenue(venue: VenueDTO !, userId: ID !): Venue
+    user: User
```
Do we want to add `user` to `type Band` as well?

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/venue/entity/Venue.java:
```diff
@@ -2,22 +2,8 @@
-import javax.persistence.SequenceGenerator;
-import javax.persistence.Table;
+
+import javax.persistence.*;
```
Please change IDE setting to avoid wildcard imports

JakubKowalski1997 committed:
User contains Venue and Band entities
Venue and Band know nothing about user
User mutation logic is now separated from what UserService should do

JakubKowalski1997 committed:
add postman scripts

JakubKowalski1997 committed:
add postman scripts

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/band/boundary/dto/BandDTO.java:
```diff
@@ -101,15 +101,15 @@ public int hashCode() {
+                ", citiesIds=" + citiesIds +
+                ", musicGenres=" + musicGenres +
+                ", songName='" + songName + '\'' +
+                '}';
```
Why don't use `StringJoiner`?

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/band/boundary/dto/BandResponseDTO.java:
```diff
@@ -111,16 +110,16 @@ public int hashCode() {
+                ", cities=" + cities +
+                ", musicGenres=" + musicGenres +
+                ", songName='" + songName + '\'' +
+                '}';
```
https://github.com/OrkiZMajorki/backend/pull/6/files#r502810830

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/boundary/UserConverter.java:
```diff
@@ -1,18 +1,40 @@
+
+    @NotNull
+    public static User convertToEntity(UserDTO userDTO) {
+        var user = new pl.orki.hackathon.webapp.user.entity.User();
```
We should avoid using package name when using classes

JakubKowalski1997 replied:
wtf intelij is doing with my code :p

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/boundary/UserResponseDTO.java:
```diff
@@ -1,14 +1,18 @@
     private Long id;
     private String username;
     private String email;
     private String role;
```
We should remove it, as we don't have this field in the schema anymore

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/boundary/UserConverter.java:
```diff
@@ -1,18 +1,40 @@
 import pl.orki.hackathon.webapp.user.entity.User;
+import pl.orki.hackathon.webapp.venue.boundary.dto.VenueResponseDTO;
 
 public class UserConverter {
```
Could you convert this class to a `@Component` without static methods? We should avoid using static methods, this was a temporary solution during hackathon ;)

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/control/UserService.java:
```diff
@@ -3,21 +3,41 @@
 
 import java.util.Optional;
 
+import static pl.orki.hackathon.webapp.user.boundary.UserConverter.*;
```
We should avoid wildcard imports

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/control/UserService.java:
```diff
@@ -31,4 +51,20 @@ public User createUser(User user) {
         user.setPassword(bCryptPasswordEncoder.encode(user.getPassword()));
```
I think this method should be private now and placed below `createUserBand` and `createUserVenue`. If you make this method private remove `@Transactional` annotation, because it won't change anything on private methods.

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/control/UserService.java:
```diff
@@ -31,4 +51,20 @@ public User createUser(User user) {
         return userRepository.save(user);
     }
+
+    public UserResponseDTO createUserBand(UserDTO userDTO, BandDTO bandDTO) {
```
Add `@Transactional` annotation

Pawlllosss replied:
I was thinking about returning entity instead of dto form service, and converting it withing `UserMutation`, but it's up to you ;)

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/control/UserService.java:
```diff
@@ -31,4 +51,20 @@ public User createUser(User user) {
+        return convertToResponseDTOWithBand(user, bandConverter.convertToDTO(user.getBand()));
+    }
+
+    public UserResponseDTO createUserVenue(UserDTO userDTO, VenueDTO venueDTO) {
```
https://github.com/OrkiZMajorki/backend/pull/6/files#r502811676

Pawlllosss commented on src/main/resources/graphql/schema/user.graphqls:
```diff
@@ -15,12 +16,14 @@ input UserLoginDTO {
-    role: String !
     email: String !
+    band: BandDTO
+    venue: VenueDTO
```
I don't think these fields are needed. `input UserDTO` is used only to describe data that we _send_, we already provide `BandDTO/VenueDTO` as a second argument for `createUserBand/createUserVenue`

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/boundary/UserDTO.java:
```diff
@@ -1,14 +1,17 @@
-    private String role;
     private String email;
+    private BandDTO band;
+    private VenueDTO venue;
```
Related to: https://github.com/OrkiZMajorki/backend/pull/6/files#r502812235

JakubKowalski1997 committed:
after review

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/control/UserService.java:
```diff
@@ -3,32 +3,72 @@
         this.userRepository = userRepository;
+        this.bandService = bandService1;
+        this.bandConverter = bandConverter;
+        this.venueService = venueService1;
```
Just `venueService`

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/control/UserService.java:
```diff
@@ -3,32 +3,72 @@
+    public UserService(BCryptPasswordEncoder bCryptPasswordEncoder, UserRepository userRepository, BandService bandService, BandService bandService1, BandConverter bandConverter, VenueService venueService, VenueService venueService1, VenueConverter venueConverter, UserConverter userConverter) {
         this.bCryptPasswordEncoder = bCryptPasswordEncoder;
         this.userRepository = userRepository;
+        this.bandService = bandService1;
```
Just `bandService`

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/boundary/UserConverter.java:
```diff
@@ -1,18 +1,42 @@
     }
+
+    @NotNull
+    public static User convertToEntity(UserDTO userDTO) {
```
We can make it the member function as well

JakubKowalski1997 committed:
after review

Pawlllosss commented on src/main/java/pl/orki/hackathon/webapp/user/entity/User.java:
```diff
@@ -78,23 +94,25 @@ public boolean equals(Object o) {
+                ", band=" + band +
+                ", venue=" + venue +
+                ", email='" + email + '\'' +
+                '}';
```
I would revert that change

Pawlllosss commented:
Please remove failing test

JakubKowalski1997 committed:
after review

Pawlllosss commented:
Test is still failing

JakubKowalski1997 committed:
delete test

JakubKowalski1997 committed:
Revert "delete test"

This reverts commit 056a36b751e045a0dbbc2a15ccaff235bca7c446.

strange but i do not have errors from this cascade type anymore

Pawlllosss committed:
Merge branch 'master' into user-refactor

Pawlllosss committed:
fixed test after merge
